palette/script.js

```diff
@@ -0,0 +1,163 @@
+    });
+
+    return arr;
+}
+function drawArray(size, arr) {
+    const canvas = document.getElementsByTagName('canvas')[0];
+    const ctx = canvas.getContext('2d');
+
+    for (let i = 0; i < arr.length; i += 1) {
+        for (let j = 0; j < arr[i].length; j += 1) {
+            ctx.fillStyle = arr[i][j];
+            ctx.fillRect(j * (canvas.width / size),
+                i * (canvas.width / size), (canvas.width / size), (canvas.width / size));
+        }
+    }
+}
+const smallImg = [
+    ['#00BCD4', '#FFEB3B', '#FFEB3B', '#00BCD4'],
+    ['#FFEB3B', '#FFC107', '#FFC107', '#FFEB3B'],
+    ['#FFEB3B', '#FFC107', '#FFC107', '#FFEB3B'],
+    ['#00BCD4', '#FFEB3B', '#FFEB3B', '#00BCD4'],
+];
+function LinearFunction(x0, y0, x1, y1) {
+    const k = (y1 - y0) / (x1 - x0);
+    const b = y0 - k * x0;
+    return (x) => k * x + b;
+}
+const drawRect = function (x, y, width, height, color) {
+    const canvas = document.getElementsByTagName('canvas')[0];
+    const context = canvas.getContext('2d');
+    context.beginPath();
+    context.rect(x, y, width, height);
+    context.fillStyle = typeof color === 'object' ? `rgba(${color.join(',')})` : `#${color}`;
+    context.fill();
+};
+function Pencil(x0, y0, x1, y1, deltaX, deltaY) {
+    let f = LinearFunction(x0, y0, x1, y1);
+    const k = (y1 - y0) / (x1 - x0);
+    const curColor = document.getElementById('current_color').value;
+    if (Math.abs(k) < 1) {
+        if (x1 > x0) {
+            for (let x = x0; x < x1; x += deltaX) {
+                const y = f(x);
+                drawRect(x, y, deltaX, deltaY, curColor);
+            }
+        } else {
+            for (let x = x0; x >= x1; x -= deltaX) {
+                const y = f(x);
+                drawRect(x, y, deltaX, deltaY, curColor);
+            }
+        }
+    } else {
+        f = LinearFunction(y0, x0, y1, x1);
+        if (y1 > y0) {
+            for (let y = y0; y < y1; y += deltaY) {
+                const x = f(y);
+                drawRect(x, y, deltaX, deltaY, curColor);
+            }
+        } else {
+            for (let y = y0; y >= y1; y -= deltaY) {
+                const x = f(y);
+                drawRect(x, y, deltaX, deltaY, curColor);
+            }
+        }
+    }
+}
+
+function ColorPicker(event, arr) {
+    document.getElementById('prev_color').style.background = CurrentColor();
+    localStorage.setItem('prev_color', CurrentColor());
+    document.getElementById('current_color').style.background = arr[Math.floor(event.offsetY / 128)][Math.floor(event.offsetX / 128)];
+    localStorage.setItem('current_color', CurrentColor());
+}
+
+window.onload = () => {
+    let arr = localStorage.getItem('canvas') ? JSON.parse(localStorage.getItem('canvas')) : smallImg;
+    drawArray(4, arr);
+    const canvas = document.getElementsByTagName('canvas')[0];
+    canvas.addEventListener('click', (event) => {
+        switch (document.querySelector('.tools-block__tools_item_active').children[1].innerHTML) {
```

Comment:
to avoid getting element on each canvas click save current tool to variable on tool change and use it in switch

Reply:
use data attributes instead of innerHTML

Reply:
ok, fixed